File: frontend/lib/types/calculation.ts
/**
 * Calculation TypeScript types — mirror of backend app/api/v1/schemas/calculation.py.
 *
 * Numeric input fields (Decimal in Python) are numbers in JSON.
 * Monetary response fields from SnapshotSummaryResponse are strings (MoneyStr).
 *
 * Architecture: IMPLEMENTATION_ROADMAP.md Commit 7.2.
 */

import type { SnapshotSummary } from "@/lib/types/snapshot";

// ---------------------------------------------------------------------------
// Enums (required by CalculationRequest)
// ---------------------------------------------------------------------------

export type MortgageType = "INTEREST_ONLY" | "REPAYMENT";

export type OwnershipStructure = "INDIVIDUAL" | "LIMITED_COMPANY";

export type IncomeTaxBand = "BASIC_RATE" | "HIGHER_RATE" | "ADDITIONAL_RATE";

export type PropertyType = "RESIDENTIAL_SINGLE_LET";

export type Tenure = "FREEHOLD" | "LEASEHOLD";

export type PropertyCountry = "ENGLAND";

export type DealStatus = "DRAFT" | "ANALYSED" | "ARCHIVED";

// ---------------------------------------------------------------------------
// Request types
// ---------------------------------------------------------------------------

/**
 * POST /api/v1/calculations/ request body.
 * Mirrors CalculationRequest. Numeric fields are numbers (Decimal → JSON number).
 */
export interface CalculationRequest {
  deal_id: string;
  calculation_date?: string | null;
  // Required inputs
  purchase_price: number;
  monthly_rent: number;
  deposit_amount: number;
  mortgage_interest_rate: number;
  mortgage_term_years: number;
  mortgage_type: MortgageType;
  ownership_structure: OwnershipStructure;
  income_tax_band?: IncomeTaxBand | null;
  is_additional_dwelling: boolean;
  property_type: PropertyType;
  tenure: Tenure;
  property_country: PropertyCountry;
  postcode: string;
  // Optional inputs — absent means use config default
  void_rate_percent?: number | null;
  letting_agent_fee_percent?: number | null;
  maintenance_reserve_percent?: number | null;
  landlord_insurance_annual?: number | null;
  purchase_legal_costs?: number | null;
  refurbishment_cost?: number | null;
  annual_service_charge?: number | null;
  annual_ground_rent?: number | null;
  annual_accountancy_cost?: number | null;
  lease_years_remaining?: number | null;
}

/**
 * POST /api/v1/calculations/recalculate request body.
 * Mirrors RecalculateRequest — uses stored working inputs, no inputs in body.
 */
export interface RecalculateRequest {
  deal_id: string;
  calculation_date?: string | null;
}

/**
 * POST /api/v1/calculations/reproduce-original request body.
 * Mirrors ReproduceOriginalRequest.
 */
export interface ReproduceOriginalRequest {
  source_snapshot_id: string;
  calculation_date?: string | null;
}

// ---------------------------------------------------------------------------
// Response types
// ---------------------------------------------------------------------------

/**
 * Mirrors CalculationSuccessResponse — successful calculation result.
 * Returned by all three calculation endpoints on success (HTTP 201).
 */
export interface CalculationSuccess {
  snapshot_id: string;
  deal_status: DealStatus;
  snapshot: SnapshotSummary;
}

/** One structured field validation error from the engine. */
export interface FieldError {
  rule_code: string;
  field: string;
  message: string;
}

/**
 * Mirrors CalculationValidationFailureResponse — engine rejected inputs.
 * Returned as HTTP 422 body when engine validation fails.
 */
export interface CalculationValidationFailure {
  detail: string;
  field_errors: FieldError[];
}

File: backend/app/api/v1/schemas/calculation.py
"""
Calculation request/response Pydantic v2 schemas.

Architecture:
    APPLICATION_SERVICE_ARCHITECTURE.md Part 15.
    IMPLEMENTATION_ROADMAP.md Commit 6.2.
"""

from __future__ import annotations

import uuid
from datetime import date
from decimal import Decimal

from pydantic import BaseModel

from app.api.v1.schemas.common import FieldError
from app.api.v1.schemas.snapshot import SnapshotSummaryResponse
from app.domain.enums import (
    DealStatus,
    IncomeTaxBand,
    MortgageType,
    OwnershipStructure,
    PropertyCountry,
    PropertyType,
    Tenure,
)


class CalculationRequest(BaseModel):
    """
    Request body for POST /api/v1/calculations/.

    Required fields must be supplied. Optional fields, if absent, are resolved
    to the active configuration defaults by ConfigurationService.resolve_defaults().
    """

    deal_id: uuid.UUID
    calculation_date: date | None = None  # defaults to today in the route handler

    # Required inputs
    purchase_price: Decimal
    monthly_rent: Decimal
    deposit_amount: Decimal
    mortgage_interest_rate: Decimal
    mortgage_term_years: int
    mortgage_type: MortgageType
    ownership_structure: OwnershipStructure
    income_tax_band: IncomeTaxBand | None = None  # required for INDIVIDUAL
    is_additional_dwelling: bool
    property_type: PropertyType
    tenure: Tenure
    property_country: PropertyCountry
    postcode: str

    # Optional inputs — absent = use config default
    void_rate_percent: Decimal | None = None
    letting_agent_fee_percent: Decimal | None = None
    maintenance_reserve_percent: Decimal | None = None
    landlord_insurance_annual: Decimal | None = None
    purchase_legal_costs: Decimal | None = None
    refurbishment_cost: Decimal | None = None
    annual_service_charge: Decimal | None = None
    annual_ground_rent: Decimal | None = None
    annual_accountancy_cost: Decimal | None = None
    lease_years_remaining: int | None = None


class CalculationSuccessResponse(BaseModel):
    """
    Response for a successful calculation.

    Returns the snapshot id, the deal's new status, and a display-level
    summary of the snapshot outputs and risk flags.
    """

    snapshot_id: uuid.UUID
    deal_status: DealStatus
    snapshot: SnapshotSummaryResponse


class CalculationValidationFailureResponse(BaseModel):
    """
    Response for a calculation that fails engine validation.

    field_errors carries structured errors that the frontend maps to
    specific form fields.
    """

    detail: str = "Calculation validation failed"
    field_errors: list[FieldError]
